chain/test_chain.go

```diff
@@ -563,27 +580,49 @@ func (t *TestChain) CallContract(msg *core.Message, state *state.StateDB, additi
 	// Create a new call tracer router that incorporates any additional tracers provided just for this call, while
 	// still calling our internal tracers.
 	extendedTracerRouter := NewTestChainTracerRouter()
-	extendedTracerRouter.AddTracer(t.callTracerRouter)
+	extendedTracerRouter.AddTracer(t.callTracerRouter.NativeTracer())
 	extendedTracerRouter.AddTracers(additionalTracers...)
 
 	// Create our EVM instance.
 	evm := vm.NewEVM(blockContext, txContext, state, t.chainConfig, vm.Config{
-		//Debug:            true,
-		Tracer:           extendedTracerRouter,
+		Tracer:           extendedTracerRouter.NativeTracer().Tracer.Hooks,
 		NoBaseFee:        true,
 		ConfigExtensions: t.vmConfigExtensions,
 	})
+	// Set our block context and chain config in order for cheatcodes to override what EVM interpreter sees.
+	t.pendingBlockContext = &evm.Context
+	t.pendingBlockChainConfig = evm.ChainConfig()
 
+	tx := utils.MessageToTransaction(msg)
+	if evm.Config.Tracer != nil && evm.Config.Tracer.OnTxStart != nil {
+		evm.Config.Tracer.OnTxStart(evm.GetVMContext(), tx, msg.From)
```

Comment:
why are we explicitly calling `OnTxStart`?

Reply:
We don't call into the code that would normally invoke this hook, so we have to do it manually https://github.com/ethereum/go-ethereum/blob/f59d013e4064adbb092b323781104c07550a8bff/core/state_processor.go#L110-L117

Reply:
Part of me feels like we can get rid of the vendored code now, right? I just looked at this function and maybe we can directly call this function and let it deal with all the tracer-related logic.

Reply:
We still need the `ExecutionResult` though so maybe we do need the vendored code...

Reply:
I think the `OnTxStart`/`OnTxEnd` logic should be moved to the `vendored` package at least for `PendingBlockAddTx`. For `CallContract` I think it is fine for now. Need to reason through this a bit more.